File: examples/python/marc_to_bibframe.py
#!/usr/bin/env python3
"""BIBFRAME conversion example: Basic MARC → BIBFRAME conversion.

This example demonstrates:
- Converting a MARC record to BIBFRAME
- Accessing the RDF graph
- Serializing to different RDF formats
"""

import mrrc


def main():
    # Create a sample MARC record using the Python API
    leader = mrrc.Leader()
    leader.record_type = "a"  # language material
    leader.bibliographic_level = "m"  # monograph
    record = mrrc.Record(leader=leader)

    # Add control fields
    record.add_control_field("001", "example-001")
    record.add_control_field("008", "040520s2023    xxu           000 0 eng  ")

    # Add ISBN
    field_020 = mrrc.Field(tag="020", indicator1=" ", indicator2=" ")
    field_020.add_subfield("a", "9780123456789")
    record.add_field(field_020)

    # Add title
    field_245 = mrrc.Field(tag="245", indicator1="1", indicator2="0")
    field_245.add_subfield("a", "Introduction to MARC /")
    field_245.add_subfield("c", "by Jane Smith.")
    record.add_field(field_245)

    # Add author
    field_100 = mrrc.Field(tag="100", indicator1="1", indicator2=" ")
    field_100.add_subfield("a", "Smith, Jane,")
    field_100.add_subfield("d", "1970-")
    field_100.add_subfield("4", "aut")
    record.add_field(field_100)

    # Add publication info
    field_260 = mrrc.Field(tag="260", indicator1=" ", indicator2=" ")
    field_260.add_subfield("a", "New York :")
    field_260.add_subfield("b", "Academic Press,")
    field_260.add_subfield("c", "2023.")
    record.add_field(field_260)

    # Add subject
    field_650 = mrrc.Field(tag="650", indicator1=" ", indicator2="0")
    field_650.add_subfield("a", "MARC (Computer record format)")
    field_650.add_subfield("x", "Cataloging.")
    record.add_field(field_650)

    print("✓ Created MARC record with 6 fields\n")

    # Convert to BIBFRAME
    config = mrrc.BibframeConfig()
    graph = mrrc.marc_to_bibframe(record, config)

    print(f"✓ Converted MARC record to BIBFRAME graph")
    print(f"  Graph contains {len(graph)} triples\n")

    # Serialize to different formats
    print("=== RDF/XML Format ===")
    rdf_xml = graph.serialize("rdf-xml")
    print(rdf_xml[:500])
    print()

    print("=== N-Triples Format (first 3 triples) ===")
    ntriples = graph.serialize("ntriples")
    for i, line in enumerate(ntriples.split("\n")):
        if i >= 3:
            break
        if line:
            print(line)

    print("\n=== JSON-LD Format ===")
    jsonld = graph.serialize("jsonld")
    print(jsonld[:300])
    print()

    print("✓ BIBFRAME conversion complete!")


if __name__ == "__main__":
    main()

File: examples/marc_to_bibframe.rs
//! BIBFRAME conversion example: Basic MARC → BIBFRAME conversion.
//!
//! This example demonstrates:
//! - Converting a MARC record to BIBFRAME
//! - Accessing the RDF graph
//! - Serializing to different RDF formats

use mrrc::bibframe::{marc_to_bibframe, BibframeConfig, RdfFormat};
use mrrc::leader::Leader;
use mrrc::record::{Field, Record};

fn main() -> Result<(), Box<dyn std::error::Error>> {
    // Create a sample MARC record
    let leader = Leader {
        record_length: 1000,
        record_status: 'n',
        record_type: 'a',
        bibliographic_level: 'm',
        control_record_type: ' ',
        character_coding: 'a',
        indicator_count: 2,
        subfield_code_count: 2,
        data_base_address: 100,
        encoding_level: ' ',
        cataloging_form: 'a',
        multipart_level: ' ',
        reserved: "4500".to_string(),
    };

    let mut record = Record::new(leader);

    // Add control fields
    record.add_control_field("001".to_string(), "example-001".to_string());
    record.add_control_field(
        "008".to_string(),
        "040520s2023    xxu           000 0 eng  ".to_string(),
    );

    // Add ISBN
    let mut f020 = Field::new("020".to_string(), ' ', ' ');
    f020.add_subfield('a', "9780123456789".to_string());
    record.add_field(f020);

    // Add title
    let mut f245 = Field::new("245".to_string(), '1', '0');
    f245.add_subfield('a', "Introduction to MARC /".to_string());
    f245.add_subfield('c', "by Jane Smith.".to_string());
    record.add_field(f245);

    // Add author
    let mut f100 = Field::new("100".to_string(), '1', ' ');
    f100.add_subfield('a', "Smith, Jane,".to_string());
    f100.add_subfield('d', "1970-".to_string());
    f100.add_subfield('4', "aut".to_string());
    record.add_field(f100);

    // Add publication info
    let mut f260 = Field::new("260".to_string(), ' ', ' ');
    f260.add_subfield('a', "New York :".to_string());
    f260.add_subfield('b', "Academic Press,".to_string());
    f260.add_subfield('c', "2023.".to_string());
    record.add_field(f260);

    // Add subject
    let mut f650 = Field::new("650".to_string(), ' ', '0');
    f650.add_subfield('a', "MARC (Computer record format)".to_string());
    f650.add_subfield('x', "Cataloging.".to_string());
    record.add_field(f650);

    // Convert to BIBFRAME with default configuration
    let config = BibframeConfig::default();
    let graph = marc_to_bibframe(&record, &config);

    println!("✓ Converted MARC record to BIBFRAME graph");
    println!("  Graph contains {} triples\n", graph.len());

    // Serialize to different formats
    println!("=== RDF/XML Format ===");
    let rdf_xml = graph.serialize(RdfFormat::RdfXml)?;
    println!("{}\n", &rdf_xml[..std::cmp::min(500, rdf_xml.len())]);

    println!("=== N-Triples Format (first 3 triples) ===");
    let ntriples = graph.serialize(RdfFormat::NTriples)?;
    for (i, line) in ntriples.lines().enumerate() {
        if i >= 3 {
            break;
        }
        if !line.is_empty() {
            println!("{line}");
        }
    }

    println!("\n=== JSON-LD Format ===");
    let jsonld = graph.serialize(RdfFormat::JsonLd)?;
    println!("{}\n", &jsonld[..std::cmp::min(300, jsonld.len())]);

    println!("✓ BIBFRAME conversion complete!");

    Ok(())
}
